okhttp/src/main/java/okhttp3/internal/cache/CacheInterceptor.kt

```diff
@@ -0,0 +1,267 @@
+ *  this work for additional information regarding copyright ownership.
+ *  The ASF licenses this file to You under the Apache License, Version 2.0
+ *  (the "License"); you may not use this file except in compliance with
+ *  the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package okhttp3.internal.cache
+
+import java.io.IOException
+import okhttp3.Headers
+import okhttp3.Interceptor
+import okhttp3.Protocol
+import okhttp3.Response
+import okhttp3.internal.Util
+import okhttp3.internal.http.ExchangeCodec
+import okhttp3.internal.http.HttpHeaders
+import okhttp3.internal.http.HttpMethod
+import okhttp3.internal.http.RealResponseBody
+import okio.Buffer
+import okio.Source
+import okio.Timeout
+import java.net.HttpURLConnection.HTTP_NOT_MODIFIED
+import java.util.concurrent.TimeUnit.MILLISECONDS
+import okhttp3.internal.Util.closeQuietly
+import okhttp3.internal.Util.discard
+import okhttp3.internal.addHeaderLenient
+import okio.buffer
+import java.net.HttpURLConnection.HTTP_GATEWAY_TIMEOUT
+
+/** Serves requests from the cache and writes responses to the cache.  */
+class CacheInterceptor(internal val cache: InternalCache?) : Interceptor {
+
+  @Throws(IOException::class)
+  override fun intercept(chain: Interceptor.Chain): Response {
+    val cacheCandidate = cache?.get(chain.request())
+
+    val now = System.currentTimeMillis()
+
+    val strategy = CacheStrategy.Factory(now, chain.request(), cacheCandidate).get()
+    val networkRequest = strategy.networkRequest
+    val cacheResponse = strategy.cacheResponse
+
+    if (cache != null) {
+      cache.trackResponse(strategy)
+    }
+
+    if (cacheCandidate != null && cacheResponse == null) {
+      // The cache candidate wasn't applicable. Close it.
+      closeQuietly(cacheCandidate.body())
+    }
+
+    // If we're forbidden from using the network and the cache is insufficient, fail.
+    if (networkRequest == null && cacheResponse == null) {
+      return Response.Builder()
+          .request(chain.request())
+          .protocol(Protocol.HTTP_1_1)
+          .code(HTTP_GATEWAY_TIMEOUT)
+          .message("Unsatisfiable Request (only-if-cached)")
+          .body(Util.EMPTY_RESPONSE)
+          .sentRequestAtMillis(-1L)
+          .receivedResponseAtMillis(System.currentTimeMillis())
+          .build()
+    }
+
+    // If we don't need the network, we're done.
+    if (networkRequest == null) {
+      return cacheResponse!!.newBuilder()
+          .cacheResponse(stripBody(cacheResponse))
+          .build()
+    }
+
+    var networkResponse: Response? = null
```

Comment:
Can you declare this as

```
var networkResponse: Response
```

don’t declare as nullable, don’t assign to null

Reply:
> Can you declare this as
> 
> ```
> var networkResponse: Response
> ```
> 
> don’t declare as nullable, don’t assign to null

I tried to declare `networkResponse:Response `means
`networkResponse` will never be null,
But actually we have a `networkResponse == null `scene:
In the finally statement:
```
// If we're crashing on I/O or otherwise, don't leak the cache body.
If (networkResponse == null && cacheCandidate != null) {
  closeQuietly(cacheCandidate.body())
}
```
After modification, it becomes:
```
If (cacheCandidate != null){
  closeQuietly(cacheCandidate.body())
}
```

When `networkResponse != null && cacheCandidate != null`,
The original logic does not `closeQuietly`,
After modification, it will `closeQuietly`, causing `networkResponse.body.bodySource.closed = true`

This causes `assertThat(response2.body().string()).isEqualTo("A")` be failed in the `conditionalHitUpdatesCache()` method `in CacheTest`.

```
java.lang.IllegalStateException: closed

At okio.RealBufferedSource.select(RealBufferedSource.kt:83)
At okhttp3.internal.Util.bomAwareCharset(Util.java:443)
At okhttp3.ResponseBody.string(ResponseBody.kt:170)
At okhttp3.CacheTest.conditionalHitUpdatesCache(CacheTest.java:1925)
```

Reply:
Cool!